src/clib/pio_lists.c

```diff
@@ -250,29 +298,46 @@ int pio_num_iosystem(int *niosysid)
 }
 
 /** 
- * Add an iodesc.
+ * Add an iodesc to a global list.
+ * This function guarantees that iodescs (id of the
+ * iodescs) are unique across the comm provided
  *
  * @param io_desc_t pointer to data to add to list.
+ * @param comm MPI Communicator across which the iosystems
+ * need to be unique
  * @returns the ioid of the newly added iodesc.
- * @author Jim Edwards
  */
-int pio_add_to_iodesc_list(io_desc_t *iodesc)
+int pio_add_to_iodesc_list(io_desc_t *iodesc, MPI_Comm comm)
 {
-    io_desc_t *ciodesc;
-    int imax = 512;
-
+    /* Using an arbitrary start id for iodesc ids helps
+     * in debugging, to distinguish between ids assigned
+     * to different structures in the code
+     */
+    static int pio_iodesc_next_id = PIO_IODESC_START_ID;
+    io_desc_t *ciodesc = pio_iodesc_list;
+
+    if(comm != MPI_COMM_NULL)
+    {
+        int tmp_id = pio_iodesc_next_id;
+        int mpierr = MPI_Allreduce(&tmp_id, &pio_iodesc_next_id, 1, MPI_INT, MPI_MAX, comm);
```

Comment:
Not sure why this allreduce is needed, if it is. Doesn't all tasks already have the same value for PIO_IODESC_START_ID?